File: core/domain/src/main/java/com/ChronosFlow/VBCR/core/domain/planner/PlannerCommand.kt
package com.ChronosFlow.VBCR.core.domain.planner

import com.ChronosFlow.VBCR.core.domain.model.TimeBlock

sealed interface PlannerCommand {
    val id: String
    val label: String

    suspend fun execute(service: PlannerService): PlannerOperationResult
    suspend fun undo(service: PlannerService): PlannerOperationResult
}

data class CreateTimeBlockCommand(
    override val id: String,
    private val block: TimeBlock
) : PlannerCommand {
    override val label: String = "Create block"

    private var created = false

    override suspend fun execute(service: PlannerService): PlannerOperationResult {
        val result = service.createBlock(block)
        created = result is PlannerOperationResult.Applied
        return result
    }

    override suspend fun undo(service: PlannerService): PlannerOperationResult {
        return if (created) {
            created = false
            service.deleteBlockById(block.id)
        } else {
            PlannerOperationResult.Rejected("Nothing to undo for create", block.id)
        }
    }
}

data class MoveTimeBlockCommand(
    override val id: String,
    private val blockId: String,
    private val targetStartMinute: Int,
    private val originalStartMinute: Int
) : PlannerCommand {
    override val label: String = "Move block"

    override suspend fun execute(service: PlannerService): PlannerOperationResult {
        return service.moveBlock(blockId, targetStartMinute)
    }

    override suspend fun undo(service: PlannerService): PlannerOperationResult {
        return service.moveBlock(blockId, originalStartMinute)
    }
}

data class ResizeTimeBlockCommand(
    override val id: String,
    private val blockId: String,
    private val targetDurationMinutes: Int,
    private val originalDurationMinutes: Int
) : PlannerCommand {
    override val label: String = "Resize block"

    override suspend fun execute(service: PlannerService): PlannerOperationResult {
        return service.resizeBlock(blockId, targetDurationMinutes)
    }

    override suspend fun undo(service: PlannerService): PlannerOperationResult {
        return service.resizeBlock(blockId, originalDurationMinutes)
    }
}

data class DeleteTimeBlockCommand(
    override val id: String,
    private val blockId: String,
    private val blockSnapshot: TimeBlock
) : PlannerCommand {
    override val label: String = "Delete block"
    private var snapshotRemoved = false

    override suspend fun execute(service: PlannerService): PlannerOperationResult {
        val result = service.deleteBlock(blockSnapshot)
        snapshotRemoved = result is PlannerOperationResult.Applied
        return result
    }

    override suspend fun undo(service: PlannerService): PlannerOperationResult {
        return if (snapshotRemoved) {
            snapshotRemoved = false
            service.createBlock(blockSnapshot)
        } else {
            PlannerOperationResult.Rejected("Nothing to undo for delete", blockId)
        }
    }
}

/** One block's repositioning within a [ResolveConflictsCommand] batch. */
data class BlockStartChange(
    val blockId: String,
    val originalStartMinute: Int,
    val newStartMinute: Int
)

/**
 * Undoable wrapper for a batch of conflict-repair relocations. The moves are already applied by
 * [PlannerService.resolveConflicts]; this command exists so the whole repair undoes/redoes as ONE
 * action. Undo/redo use the forced [PlannerService.setBlockStart] (no placement validation) because
 * undo deliberately recreates the prior — overlapping — layout.
 */
data class ResolveConflictsCommand(
    override val id: String,
    private val changes: List<BlockStartChange>
) : PlannerCommand {
    override val label: String = "Repair schedule"

    override suspend fun execute(service: PlannerService): PlannerOperationResult {
        changes.forEach { service.setBlockStart(it.blockId, it.newStartMinute) }
        return PlannerOperationResult.Applied(
            "Schedule repaired", changes.firstOrNull()?.blockId.orEmpty(), null, changes.map { it.blockId }
        )
    }

    override suspend fun undo(service: PlannerService): PlannerOperationResult {
        changes.forEach { service.setBlockStart(it.blockId, it.originalStartMinute) }
        return PlannerOperationResult.Applied(
            "Repair undone", changes.firstOrNull()?.blockId.orEmpty(), null, changes.map { it.blockId }
        )
    }
}

File: ios/ChronosCore/Sources/ChronosCore/PlannerCommand.swift
import Foundation

// PlannerCommand — pure, Foundation-only, value-typed port of `core/domain/planner/PlannerCommand.kt`.
//
// Android's PlannerCommand is a sealed interface whose execute/undo run *through a PlannerService*
// (suspend, hits Room). ChronosCore is pure and has no service, so each command here applies and
// inverts directly over an immutable `[PlannerBlock]` (the day's blocks), returning a new array —
// the same shape as a transactional reducer. The iOS app runs a command, then mirrors the resulting
// array into its SwiftData store. This keeps the whole undo/redo audit trail portable and testable.
//
// Semantics mirror Android exactly:
//  • Create:  apply = append the block;            inverse = remove it by id.
//  • Delete:  apply = remove the snapshot by id;   inverse = re-insert the snapshot.
//  • Move:    apply = set start = target;          inverse = set start = original.
//  • Resize:  apply = set duration = target;       inverse = set duration = original.
//  • ResolveConflicts: apply = force each block's start to its new value (no validation, mirrors
//    PlannerService.setBlockStart); inverse = force each back to its original — deliberately
//    recreating the prior, possibly overlapping, layout, as ONE undoable batch.
//
// Move/Resize/ResolveConflicts carry the *original* values so they invert losslessly without a
// service round-trip. Everything is `Sendable` and value-typed.

/// One block's repositioning within a `resolveConflicts` batch. Port of `BlockStartChange`.
public struct BlockStartChange: Sendable, Equatable {
    public let blockId: String
    public let originalStartMinute: Int
    public let newStartMinute: Int
    public init(blockId: String, originalStartMinute: Int, newStartMinute: Int) {
        self.blockId = blockId
        self.originalStartMinute = originalStartMinute
        self.newStartMinute = newStartMinute
    }
}

/// A value-typed, undoable planner edit. Mirrors the `PlannerCommand` sealed interface — the cases
/// are `CreateTimeBlockCommand`, `MoveTimeBlockCommand`, `ResizeTimeBlockCommand`,
/// `DeleteTimeBlockCommand`, and `ResolveConflictsCommand`. `apply`/`inverse` are pure functions of
/// the day's blocks: they return a *new* `[PlannerBlock]` and never mutate in place.
public enum PlannerCommand: Sendable, Equatable, Identifiable {
    /// Append a freshly-built block to the day.
    case create(id: String, block: PlannerBlock)
    /// Re-anchor a block's start. Carries the original start so it inverts losslessly.
    case move(id: String, blockId: String, targetStartMinute: Int, originalStartMinute: Int)
    /// Re-size a block's duration. Carries the original duration so it inverts losslessly.
    case resize(id: String, blockId: String, targetDurationMinutes: Int, originalDurationMinutes: Int)
    /// Remove a block, keeping a full snapshot so undo restores it exactly.
    case delete(id: String, blockSnapshot: PlannerBlock)
    /// A batch of conflict-repair relocations that undoes/redoes as one action.
    case resolveConflicts(id: String, changes: [BlockStartChange])

    /// Stable command id (Android's `PlannerCommand.id`). Used as the command-history key.
    public var id: String {
        switch self {
        case let .create(id, _),
             let .move(id, _, _, _),
             let .resize(id, _, _, _),
             let .delete(id, _),
             let .resolveConflicts(id, _):
            return id
        }
    }

    /// Human-readable label (Android's `PlannerCommand.label`).
    public var label: String {
        switch self {
        case .create: return "Create block"
        case .move: return "Move block"
        case .resize: return "Resize block"
        case .delete: return "Delete block"
        case .resolveConflicts: return "Repair schedule"
        }
    }

    /// Run the command forward, returning the updated day. Pure — no wall clock, no service.
    public func apply(to blocks: [PlannerBlock]) -> [PlannerBlock] {
        switch self {
        case let .create(_, block):
            return PlannerCommand.insert(block, into: blocks)
        case let .move(_, blockId, target, _):
            return PlannerCommand.setStart(blockId, to: target, in: blocks)
        case let .resize(_, blockId, target, _):
            return PlannerCommand.setDuration(blockId, to: target, in: blocks)
        case let .delete(_, snapshot):
            return PlannerCommand.remove(snapshot.id, from: blocks)
        case let .resolveConflicts(_, changes):
            return PlannerCommand.applyStarts(changes.map { ($0.blockId, $0.newStartMinute) }, in: blocks)
        }
    }

    /// Run the command backward, restoring the prior day. Inverse of `apply`.
    public func inverse(on blocks: [PlannerBlock]) -> [PlannerBlock] {
        switch self {
        case let .create(_, block):
            return PlannerCommand.remove(block.id, from: blocks)
        case let .move(_, blockId, _, original):
            return PlannerCommand.setStart(blockId, to: original, in: blocks)
        case let .resize(_, blockId, _, original):
            return PlannerCommand.setDuration(blockId, to: original, in: blocks)
        case let .delete(_, snapshot):
            return PlannerCommand.insert(snapshot, into: blocks)
        case let .resolveConflicts(_, changes):
            return PlannerCommand.applyStarts(changes.map { ($0.blockId, $0.originalStartMinute) }, in: blocks)
        }
    }

    // MARK: - Pure reducers

    /// Append unless a block with the same id already exists (idempotent re-insert on undo).
    private static func insert(_ block: PlannerBlock, into blocks: [PlannerBlock]) -> [PlannerBlock] {
        guard !blocks.contains(where: { $0.id == block.id }) else { return blocks }
        return blocks + [block]
    }

    private static func remove(_ id: String, from blocks: [PlannerBlock]) -> [PlannerBlock] {
        blocks.filter { $0.id != id }
    }

    private static func setStart(_ id: String, to start: Int, in blocks: [PlannerBlock]) -> [PlannerBlock] {
        blocks.map { block in
            guard block.id == id else { return block }
            var updated = block
            updated.startMinuteOfDay = start
            return updated
        }
    }

    private static func setDuration(_ id: String, to duration: Int, in blocks: [PlannerBlock]) -> [PlannerBlock] {
        blocks.map { block in
            guard block.id == id else { return block }
            var updated = block
            updated.durationMinutes = duration
            return updated
        }
    }

    private static func applyStarts(_ starts: [(String, Int)], in blocks: [PlannerBlock]) -> [PlannerBlock] {
        let byId = Dictionary(starts, uniquingKeysWith: { _, last in last })
        return blocks.map { block in
            guard let start = byId[block.id] else { return block }
            var updated = block
            updated.startMinuteOfDay = start
            return updated
        }
    }
}
